### DISCUSSION TITLE
Proposal: Mark "default", as not-intentional, using new Keyword "vague"

### DISCUSSION THREAD

**Discussion by ghost** [dotnet/csharplang#6974] [Language Ideas] [answered]
Currently, **there is no way to denote if a default/null value was intentionally assigned**.

It has to be standardized within each organization what these values represent, this leads to:

+ Lack of a simple way to express intention
+ Less separation of logic between modules
+ Misuse of code from external-organizations

All of which is not ideal, for an object-oriented language in use by commercial and open-source consumers.

**A new keyword "vague" should be added to the C#-language, which sets the variable to its "default" but *also* marks it as an unintentional.**

This allows for more precise error-checking, and for open use of default-values in methods.
Let’s see an example in action:

```csharp

public readonly struct GirlScoutInventory
{
    public readonly int Cookies;

    // Cookies can also be set to '0', which is default(int)!!!
    public GirlScoutInventory(int cookies)
    {
        Cookies = cookies;
    }
}

public void IsValueImportant(GirlScoutInventory inventory)
{
    // In the event cookies was set to '0' this will activate.
    // Compiler cannot verify if this value was intentional or not.
    // This means either input-checking, or default, can’t exist.
    // Another way to solve this is with an internal bool, but that is convoluted.
    if (inventory == default)…

    // This should NEVER be an accepted-value. Therefore it’s always invalid.
    // This standard currently does not exist in C#.
    if (inventory == vague || inventory is vague)…
}
```

Some may argue that this is the point of optionals, or in C# System.Nullable<T>, the '?'-operator. But I would argue that this solution does not fit the bill either. Let’s see another example:

```csharp

public void IsValueReallyImportant(GirlScoutInventory? inventory)
{
    // Problem solved right? Not really.
    // Inventory being null only seems erroneous because you KNOW it’s a struct.
    // This requires an unconditioned reader to refer to the object-in-question.
    // Furthermore, this relegates null to being for errors rather than logic.
    // That means for generic-values, or classes, this solution wouldn’t work.
    if (inventory == null || inventory is null)…


    // This should NEVER be an accepted-value. Therefore it’s always invalid.
    // Reader's don’t care about what the object is. This is always erroneous.
    // This standard currently does not exist in C#.
    if (inventory == vague || inventory is vague)…
}
```

The keyword doesn’t necessarily have to be "vague", but it seems to fit well.
Please leave improvements, feedback, and thoughts on the proposal below. It is appreciated.

**Comment by HaloFour** [comment, marked as answer]
A new keyword wouldn't help here.  In this case every single bit in the `int` datatype (or `GirlScoutInventory` for that matter) is dedicated to the integral value and its sign.  There is nowhere to pass any out-of-bound information to convey intention.  It's impossible for the compiler to discern that based on the value that was passed to the method.

For `GirlScoutInventory` and other similar types you do have the option to add another field which can be used to convey intention:

```cs
public readonly struct GirlScoutInventory
{
    public readonly int Cookies;
    public readonly bool IsSet;

    public GirlScoutInventory(int cookies)
    {
        Cookies = cookies;
        IsSet = true;
    }
}

public void IsValueImportant(GirlScoutInventory inventory)
{
    // if the method was called with `default(GirlScoutInventory)` then `IsSet` would be `false`
    if (inventory.IsSet) { ... }
}
```

**Reply by ghost** [reply]
Are structs in C# compiled the same way they are in C?

My understanding is that structs also have an object-header because it’s in managed-runtime. If so, would it not be possible to fit one-bit into the header?

If they are compiled like C, then I understand what you are saying - it just wouldn’t be possible.

___

I also agree with the bool solution, I currently do something to that effect to solve this issue. But I feel having a level of intentionality in the language itself, if possible, would be the best solution.

It standardizes what is considered "unknown", which is a really massive-step for error-checking and method-flexibility.

**Reply by HaloFour** [reply]
Structs do not have a managed header unless they are boxed on the heap (assigned to an `object`).  Otherwise they work very similarly to how they work in C.  That `GirlScoutInventory` struct is the same size as an `int`, which is 32-bits.

See: https://learn.microsoft.com/en-us/dotnet/csharp/programming-guide/types/boxing-and-unboxing#boxing

**Reply by ghost** [reply]
Yeah, then it’s just not possible. Thanks for the clarification.
